README.md
[![crates.io](https://img.shields.io/crates/v/s_curve.svg)](https://crates.io/crates/s_curve)
[![GitHub Workflow Status](https://img.shields.io/github/actions/workflow/status/marcbone/s_curve/rust.yml)](https://github.com/marcbone/s_curve/actions)
[![crates.io](https://img.shields.io/crates/l/s_curve.svg)](https://crates.io/crates/s_curve)
[![crates.io](https://img.shields.io/crates/d/s_curve.svg)](https://crates.io/crates/s_curve)
[![docs.rs](https://docs.rs/s_curve/badge.svg)](https://docs.rs/s_curve)
[![codecov](https://codecov.io/gh/marcbone/s_curve/branch/main/graph/badge.svg?token=9LB5SI9B1E)](https://codecov.io/gh/marcbone/s_curve)



A library to compute S-Curve trajectories. It can be used to generate motion profiles for robotics.

## What is an S-Curve?

An S-Curve is a trajectory which is constrained to maximum jerk, acceleration and velocity.
An S-Curve consists of 7 phases:
 * constant maximum jerk until the desired acceleration is reached
 * constant maximum acceleration phase
 * constant minimum jerk until the desired velocity is reached with an acceleration of zero
 * constant velocity phase
 * constant minimum jerk until the minimum acceleration is reached
 * constant minimum acceleration phase
 * constant maximum jerk until the acceleration is zero and the desired position and end velocity is reached
 
 In the picture below you can see an S-Curve Profile which goes from Position 0 to position 10 within 5.5 seconds with a start and end velocity of 0
, a maximum jerk of 3, a maximum acceleration of 2 and a maximum velocity of 3.
[![image](http://i.imgur.com/BQPhS8n.png)](http://i.imgur.com/BQPhS8n.png)
## Example
  ```rust
  use s_curve::*;
  let constraints = SCurveConstraints {
              max_jerk: 3.,
              max_acceleration: 2.0,
              max_velocity: 3.};
  let  start_conditions = SCurveStartConditions {
      q0: 0., // start position
      q1: 10., // end position
      v0: 0., // start velocity
      v1: 0. // end velocity
  };
let input  =  SCurveInput{constraints, start_conditions};
let (params,s_curve) = s_curve_generator( & input,Derivative::Velocity);
for i in 0..101 {
      println!("{}", s_curve(i as f64 * params.time_intervals.total_duration() / 100.));
  }
  ```
#### License
Copyright (c) 2020 Marco Boneberger
Licensed under either of [Apache License, Version 2.0](LICENSE-APACHE) or [MIT license](LICENSE-MIT) at your option. 



CHANGELOG.md
# Changelog

## [0.1.7] - 2023-05-02

### Added

- Changelog

### Fixed

- Library handles negative directions properly now
- Fixed badge in README

## [0.1.6] - 2021-05-01

### Changed

- Calculation of the time intervals is now much faster

## [0.1.5] - 2020-12-10

### Changed

- Replaced powf with powi in internal implementation

## [0.1.4] - 2020-10-12
- Initial release


examples/plot.rs
//Copyright (c) 2020 Marco Boneberger
use gnuplot::Coordinate::Graph;
use gnuplot::{AxesCommon, Caption, Figure};
use s_curve::*;

fn main() {
    let constraints = SCurveConstraints {
        max_jerk: 3.,
        max_acceleration: 2.0,
        max_velocity: 3.,
    };
    let start_conditions = SCurveStartConditions {
        q0: 0.,
        q1: 10.,
        v0: 0.,
        v1: 0.,
    };
    let input = SCurveInput {
        constraints,
        start_conditions,
    };
    let (params, s_curve) = s_curve_generator(&input, Derivative::Velocity);
    let mut x: Vec<f64> = Vec::new();
    let mut y: Vec<f64> = Vec::new();
    for i in 0..1001 {
        x.push(i as f64 * params.time_intervals.total_duration() / 1000.);
        y.push(s_curve(
            i as f64 * params.time_intervals.total_duration() / 1000.,
        ));
    }
    let mut fg = Figure::new();
    fg.axes2d()
        .set_title("S-Curve Velocity Motion Profile", &[])
        .set_legend(Graph(0.5), Graph(0.9), &[], &[])
        .set_x_label("time in seconds", &[])
        .set_y_label("velocity in m/s", &[])
        .lines(x.clone(), y.clone(), &[Caption("Velocity")]);
    fg.show().unwrap();
}


examples/plotall.rs
//Copyright (c) 2020 Marco Boneberger
use gnuplot::Coordinate::Graph;
use gnuplot::{AxesCommon, Caption, Figure};
use s_curve::*;

fn main() {
    let constraints = SCurveConstraints {
        max_jerk: 3.,
        max_acceleration: 2.0,
        max_velocity: 3.,
    };
    let start_conditions = SCurveStartConditions {
        q0: 0.,
        q1: 10.,
        v0: -1.,
        v1: -1.,
    };
    let input = SCurveInput {
        constraints,
        start_conditions,
    };
    // jerk
    let (paramsjer, s_curvejer) = s_curve_generator(&input, Derivative::Jerk);
    let mut xjer: Vec<f64> = Vec::new();
    let mut yjer: Vec<f64> = Vec::new();
    for i in 0..1001 {
        xjer.push(i as f64 * paramsjer.time_intervals.total_duration() / 1000.);
        yjer.push(s_curvejer(
            i as f64 * paramsjer.time_intervals.total_duration() / 1000.,
        ));
    }
    // pos
    let (paramspos, s_curvepos) = s_curve_generator(&input, Derivative::Position);
    let mut xpos: Vec<f64> = Vec::new();
    let mut ypos: Vec<f64> = Vec::new();
    for i in 0..1001 {
        xpos.push(i as f64 * paramspos.time_intervals.total_duration() / 1000.);
        ypos.push(s_curvepos(
            i as f64 * paramspos.time_intervals.total_duration() / 1000.,
        ));
    }
    // vel
    let (paramsvel, s_curvevel) = s_curve_generator(&input, Derivative::Velocity);
    let mut xvel: Vec<f64> = Vec::new();
    let mut yvel: Vec<f64> = Vec::new();
    for i in 0..1001 {
        xvel.push(i as f64 * paramsvel.time_intervals.total_duration() / 1000.);
        yvel.push(s_curvevel(
            i as f64 * paramsvel.time_intervals.total_duration() / 1000.,
        ));
    }
    // acc
    let (paramsacc, s_curveacc) = s_curve_generator(&input, Derivative::Acceleration);
    let mut xacc: Vec<f64> = Vec::new();
    let mut yacc: Vec<f64> = Vec::new();
    for i in 0..1001 {
        xacc.push(i as f64 * paramsacc.time_intervals.total_duration() / 1000.);
        yacc.push(s_curveacc(
            i as f64 * paramsacc.time_intervals.total_duration() / 1000.,
        ));
    }
    let mut fg = Figure::new();
    fg.axes2d()
        .set_title("S-Curve Velocity Motion Profile", &[])
        .set_legend(Graph(0.5), Graph(0.9), &[], &[])
        .set_x_label("time in seconds", &[])
        .set_y_label("Position derivatives m, m/s, m/s², m/s³", &[])
        .lines(xpos.clone(), ypos.clone(), &[Caption("Position")])
        .lines(xvel.clone(), yvel.clone(), &[Caption("Velocity")])
        .lines(xacc.clone(), yacc.clone(), &[Caption("Acceleration")])
        .lines(xjer.clone(), yjer.clone(), &[Caption("Jerk")]);
    fg.show().unwrap();
}
